File: python-library-typed/graph/kruskal.py
from typing import List, TypeVar, Generic, Tuple
T = TypeVar('T', int, float)


class Edge(Generic[T]):
    def __init__(self, dst: int, weight: T) -> None:
        self.dst: int = dst
        self.weight: T = weight

    def __lt__(self, e: 'Edge') -> bool:
        return self.weight > e.weight


class Graph(Generic[T]):
    def __init__(self, V: int) -> None:
        self.V: int = V
        self.E: List[List[Edge[T]]] = [[] for _ in range(V)]
        self.edge_size: int = 0

    def add_edge(self, src: int, dst: int, weight: T) -> None:
        self.E[src].append(Edge(dst, weight))
        self.edge_size += 1


class UnionFindTree:
    def __init__(self, N: int) -> None:
        self.par: List[int] = list(range(N))  # parent
        self.rank: List[int] = [0] * N  # depth of tree

    def find(self, x: int) -> int:
        if self.par[x] == x:
            return x
        else:
            self.par[x] = self.find(self.par[x])
            return self.par[x]

    def unite(self, x: int, y: int) -> None:
        x, y = self.find(x), self.find(y)
        if x == y:
            return
        if self.rank[x] < self.rank[y]:
            self.par[x] = y
        else:
            self.par[y] = x
            if self.rank[x] == self.rank[y]:
                self.rank[x] += 1

    def same(self, x: int, y: int) -> bool:
        return self.find(x) == self.find(y)


class MinimumSpanningTree(Generic[T]):
    def __init__(self, G: Graph) -> None:
        self.G = G

    def run(self) -> None:
        edges: List[Tuple[T, int, int]] = []
        for v1 in range(self.G.V):
            for e in self.G.E[v1]:
                v2: int = e.dst
                if v1 < v2:
                    edges.append((e.weight, v1, v2))
        edges.sort(reverse=True)

        self.mincost: T = 0
        self.minSpanningTree: Graph[T] = Graph(self.G.V)
        uf: UnionFindTree = UnionFindTree(self.G.V)

        while self.minSpanningTree.edge_size < self.G.V - 1:
            cost, v1, v2 = edges.pop()
            if not uf.same(v1, v2):
                self.mincost += cost
                uf.unite(v1, v2)
                self.minSpanningTree.add_edge(v1, v2, cost)
                self.minSpanningTree.add_edge(v2, v1, cost)

    def minCost(self) -> T:
        return self.mincost

    def getMinSpanningTree(self) -> Graph[T]:
        return self.minSpanningTree

File: c++-library/kruskal.cpp
#include <vector>

template <typename T>
struct Edge {
  int src, dst;
  T weight;

  Edge(int src, int dst, T weight) : src(src), dst(dst), weight(weight) {}

  bool operator<(const Edge<T> &e) const { return weight < e.weight; }
};

template <typename T>
struct Graph {
  int V;
  std::vector<Edge<T>> E;

  Graph(int V) : V(V) {}

  void add_edge(int src, int dst, T weight) {
    E.emplace_back(src, dst, weight);
  }
};

struct DisjointSet {
  std::vector<int> parent, rank;

  DisjointSet(int N) {
    parent.resize(N), rank.resize(N, 0);
    for (int i = 0; i < N; i++) parent[i] = i;
  }

  int find(int x) {
    if (parent[x] == x) {
      return x;
    } else {
      return parent[x] = find(parent[x]);
    }
  }

  void unite(int x, int y) {
    x = find(x), y = find(y);
    if (x == y) return;
    if (rank[x] < rank[y]) {
      parent[x] = y;
    } else {
      parent[y] = x;
      if (rank[x] == rank[y]) rank[x]++;
    }
  }

  bool same(int x, int y) { return find(x) == find(y); }
};

template <typename T>
struct MinimumSpanningTree {
  Graph<T> g;
  std::vector<Edge<T>> mst;
  T mincost;

  MinimumSpanningTree(const Graph<T> &g) : g(g) {}

  T run() {
    mincost = 0;
    sort(g.E.begin(), g.E.end());

    DisjointSet uf(g.V);

    for (int i = 0, edge_count = g.V - 1; edge_count; i++) {
      assert(i < (int)g.E.size());
      Edge<T> e = g.E[i];
      if (!uf.same(e.src, e.dst)) {
        uf.unite(e.src, e.dst);
        mst.emplace_back(e);
        mincost += e.weight;
        edge_count--;
      }
    }
    return mincost;
  }
};
